controllers/tokens/agtokens_reconciler.go

```diff
@@ -0,0 +1,132 @@
+package tokens
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	dynatracev1alpha1 "github.com/Dynatrace/dynatrace-operator/api/v1alpha1"
+	"github.com/Dynatrace/dynatrace-operator/dtclient"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	SecretsName = "dynatrace-tokens"
+)
+
+type Reconciler struct {
+	client.Client
+	apiReader client.Reader
+	instance  *dynatracev1alpha1.DynaKube
+	dtc       dtclient.Client
+	log       logr.Logger
+	scheme    *runtime.Scheme
+}
+
+func NewReconciler(clt client.Client, apiReader client.Reader, scheme *runtime.Scheme, instance *dynatracev1alpha1.DynaKube, dtc dtclient.Client, log logr.Logger) *Reconciler {
+	return &Reconciler{
+		Client:    clt,
+		apiReader: apiReader,
+		scheme:    scheme,
+		instance:  instance,
+		dtc:       dtc,
+		log:       log,
+	}
+}
+
+func (r *Reconciler) Reconcile() (bool, error) {
+	if r.instance.Spec.AGTokensSecret == "" {
+		err := r.reconcileAGTokens()
+		if err != nil {
+			r.log.Error(err, "could not reconcile ag tokens")
+			return false, errors.WithStack(err)
+		}
+	}
+
+	return false, nil
+}
+
+func (r *Reconciler) reconcileAGTokens() error {
+	agTokens, err := r.getAGTokens()
+	if err != nil {
+		return fmt.Errorf("could not fetch AG tokens: %w", err)
+	}
+
+	agTokensSecret, err := r.createAGTokensSecretIfNotExists(agTokens)
+	if err != nil {
+		return fmt.Errorf("failed to create or update secret: %w", err)
+	}
+
+	return r.updateAGTokensIfOutdated(agTokensSecret, agTokens)
```

Comment:
A bit difficult to understand. I would merge those two methods as they the one calls the other anyways.

Reply:
It's a valid point but that's the pattern that's already in use. 
```
$ ag --go IfOutdated
controllers/dtpullsecret/reconciler.go
69:     return r.updatePullSecretIfOutdated(pullSecret, pullSecretData)
82:func (r *Reconciler) updatePullSecretIfOutdated(pullSecret *corev1.Secret, desiredPullSecretData map[string][]byte) error {

controllers/customproperties/reconciler.go
55:                     err = r.updateCustomPropertiesIfOutdated()
76:func (r *Reconciler) updateCustomPropertiesIfOutdated() error {

controllers/tokens/agtokens_reconciler.go
67:     return r.updateAGTokensIfOutdated(agTokensSecret, agTokens)
80:func (r *Reconciler) updateAGTokensIfOutdated(agTokens *corev1.Secret, agTokensSecretData map[string][]byte) error {

controllers/activegate/reconciler/statefulset/reconciler_test.go
168:func TestReconcile_UpdateStatefulSetIfOutdated(t *testing.T) {
174:    updated, err := r.updateStatefulSetIfOutdated(desiredSts)
183:    updated, err = r.updateStatefulSetIfOutdated(desiredSts)
191:    updated, err = r.updateStatefulSetIfOutdated(desiredSts)

controllers/activegate/reconciler/statefulset/reconciler.go
115:    updated, err := r.updateStatefulSetIfOutdated(desiredSts)
160:func (r *Reconciler) updateStatefulSetIfOutdated(desiredSts *appsv1.StatefulSet) (bool, error) {
```
I think such refactoring should be done in separate PR.